### ISSUE TITLE
Decluttering Czech Keyboard

### ISSUE THREAD

**Issue by frimdo** [Julow/Unexpected-Keyboard#349]
As a native czech speaker and user of czech qwertz keyboard, I find the `qwertz_cs.xml` layout cluttered.
There are many accents unused in czech language (ß,ô,ľ,ë,ï,ö,ü,...), greek letters and unneeded math symbols (×,⋅).

Here is image characters that I find useless (as they can be enabled using `Add keys to the keyboard` feature.
![obrazek](https://user-images.githubusercontent.com/10960486/234841563-50b41b8c-1848-4bbe-ae87-2ad5cd1519c0.png)

Also, placing of the letters is unusual compared to czech qwertz keyboard (as seen from e.g. wiki):
![wiki_qwertz](https://upload.wikimedia.org/wikipedia/commons/thumb/9/98/Windows_Czech_keyboard_layout.svg/800px-Windows_Czech_keyboard_layout.svg.png)

Here is image of proposed layout:
![obrazek](https://user-images.githubusercontent.com/10960486/234842653-14b62966-9a02-4e29-97ef-3198e3e432cf.png)
Ans here is proposed layout:

```
<?xml version="1.0" encoding="utf-8"?>
<keyboard>
  <row>
    <key                 key0="q"         key1="1"         key2="esc"       key3="+"          key4="~"      />
    <key                 key0="w"         key1="2"                          key3="|"          key4="\\"     />
    <key                 key0="e"         key1="3"         key2="é"         key3="^"          key4="ě"      />
    <key                 key0="r"         key1="4"         key2="ŕ"                           key4="ř"      />
    <key                 key0="t"         key1="5"                          key3="°"          key4="ť"      />
    <key                 key0="z"         key1="6"                                            key4="ž"      />
    <key                 key0="u"         key1="7"         key2="ú"         key3="\`"         key4="ů"      />
    <key                 key0="i"         key1="8"         key2="í"         key3="("                        />
    <key                 key0="o"         key1="9"         key2="ó"         key3=")"                        />
    <key                 key0="p"         key1="0"         key2="/"         key3="="          key4="%"      />
  </row>
  <row>
    <key shift="0.5"     key0="a"         key1="tab"       key2="á"         key3=";"                        />
    <key                 key0="s"                                                             key4="š"      />
    <key                 key0="d"                                                             key4="ď"      />
    <key                 key0="f"                                           key3="["                        />
    <key                 key0="g"                                           key3="]"                        />
    <key                 key0="h"                                                                           />
    <key                 key0="j"                                                                           />
    <key                 key0="k"                                                                           />
    <key                 key0="l"          key1="&quot;"   key2="\'"        key3="$"          key4="!"      />
  </row>
  <row>
    <key width="1.5"     key0="shift"                      key2="loc capslock"                              />
    <key                 key0="y"                          key2="ý"                                         />
    <key                 key0="x"                                           key3="\#"                       />
    <key                 key0="c"                                           key3="&amp;"      key4="č"      />
    <key                 key0="v"                                           key3="\@"                       />
    <key                 key0="b"          key1="&gt;"     key2="&lt;"      key3="{"          key4="}"      />
    <key                 key0="n"          key1="\?"       key2="."         key3=","          key4="ň"      />
    <key                 key0="m"          key1=":"        key2="*"         key3="-"          key4="_"      />
    <key width="1.5"     key0="backspace"                  key2="delete"                                    />
  </row>
</keyboard>
```

I failed to load any keyboard layout using `custom layout` feature, so there can be some kind of error.
I wanted to get some feedback befor submitting the proposed layout as pull request.

Also, in `res/xml/method.xml`, there is line
```
<subtype android:label="%s" android:languageTag="cs" android:imeSubtypeLocale="cs_CZ" android:imeSubtypeMode="keyboard" android:isAsciiCapable="true" android:imeSubtypeExtraValue="default_layout=qwertz_cs,extra_keys=a|accent_cedille|accent_circonflexe"/>
```
what is the purpose of the `extra_keys=a|accent_cedille|accent_circonflexe"`? Cedille nor circonflexe accents are not used in czech.

**Comment by Julow** [comment]
Thanks for your proposal. I am entirely in favor of simpler layouts and I'd prefer to move these rarely used symbols in other parts (modifiers, math layout, new features).

@Validbit you contributed this layout in https://github.com/Julow/Unexpected-Keyboard/pull/198, do you have an opinion ?

**Comment by Julow** [comment]
Your use case might be better by the "extra keys" feature.

Currently, if you install the 3 languages on your device, you'll get these symbols added to every layouts: `a, accent_cedille, accent_circonflexe, accent_trema, ß, €, é, ě, ŕ, ř, ť, ž, ü, ú, ů, í, ö, ó, ô, á, ä, ś, š, ď, ĺ, ľ, ý, č, ň`.
As @frimdo said, the list for Czech is wrong.

Many Slovakian letters like éýúó, etc.. could be better typed by using `accent_aigu`. This is something that I mean to implement someday: An option to choose between accented letters or deadkeys for the "extra keys".

**Comment by Validbit** [comment]
I agree, but I see as best option **preserving both** (Czech + Czech Multilingual), but the disappearing language option (here without default cs) should be fixed (if not yet) since people use it (idk how many outside of my circles)

_1. (generally for all layouts) ∙ and × are used often and should be considered to be preserved + grave accent is used almost everywhere (nowadays) in custom formatting (or commenting out code) **(should be discussed in separate topic)**_

2. Most people won't be setting this up or won't know it's there (not generally in other keyboards) + the position/layout cannot be specified (can be placed inconveniently)

4. Again, if there are both layouts to choose from I agree :)

**Comment by Julow** [comment]
@Validbit What keys from what languages do you need on your layout ?

- `∙` and `×` could be added to the `Fn` layer.
- `` ` `` and `|` are part of the ASCII punctuation used by programmers. I recommend layouts for latin languages to include them as they are sometimes used in passwords but it's not mandatory (in fact, most layouts lack some of them).
- What do you think of the layout in https://github.com/Julow/Unexpected-Keyboard/pull/380 assuming the multilingual keys are added automatically to it ?

**Comment by frimdo** [comment]
Just to be complete, the ``` ` ``` and `|` characters are included in layout proposed in #380 (bottom left character for `w` and `o` buttons respectively).

**Comment by Validbit** [comment]
I revised several Czech layouts across keyboards (both physical and software), incl. latest Unexpected and taken into account current feedback from Github and friends and merged it into latest version (fork). 

_Generally it's difficult (and imo counter-(productive/intuitive)) to make simple universal single-language layout (even Google keyboard has symbols from languages I never used with single language selected; latest Unexpected English/Universal QWERTZ layout has German characters (historically) - although compliant with AOSP)) 
Here (and in other apps) QWERTZ and QWERTY differers drastically in symbol placement while in physical layouts (utilizing ALT Gr symbols) the layouts are almost identical (QWERTZ, QWERTY)._

Nontheless I've adjusted and recreated them to accomodate for most requests/situations (most used layouts).

- Czech (minimal)
- Czech AOSP (Powerusers, Legacy) - Closest to requested changes and popular layouts
- (Czech) Multifunctional (extended default for practical usage in academic and work use in Central Europe - incl. Slovak, German letters & symbols)

------------------------------------------
For mentioned suggestions/changes:
- ∙ yes, it's now removed - it's not as commonly used 
(it is still common in elementary & high education)
- `×` should be preserved (in full layout, imho) as it may be used often by many (like `+`, `-`) to denote multiplication where `*` can cause confusion (like in markdown and other popular lanaguages/parsing)

- most symbols should be where requested (AOSP layout) with some symbols placed in common location (in AOSP/OEM and other keyboards) and `\` `~` moved by 1 key for better one-handed operation 
- For French (from what @Julow written, if the list of accents is exact) it's missing  â, ê, î, û characters (with accent circonflexe)

I re-read what @Julow said and I understand there are automatic multilingual capabilities in place but it didn't seem to work (at least with the #380 in emulator; if you meant automatic addition of individual accented characters)*, where certain characters were missing.

I see a possibilty to implement (optional) automatic multilingual layouts by adding switch (settings option) to
- Use system languages (to dynamically add language accents)
- Use system languages (to add language accents' extra keys)
- Use predefined layout(s)\*\* 
where the first option (dynamic) could prioritize accent placement 
(to intuitively make characters with e.g. circonflex ←↓, acute+umlaut ↓→, grave →↑), if not blocked by symbols)

*(in my testing it didn't add most characters in Czech, French, Turkey... if any)
\*\*(Secondary layout is sufficient for most cases, even though it could be and array instead of iterating over (switching between) two) (i see more issues have been already opened for this)

------------------------------------------
Additional changes made:
- Fn + × = ∙ (unset before)
- Fn + § = ¶ (unset before; means the same thing)
- Updated and Translated new strings	

PS: 
- is making Fn + / = ÷ safe (not to break shortcuts or something internal)?
- is there possibility to hide extra keys (defined per language) by default (or at least adding string in-app (after install) that ~"if the keyboard seems cluttered for you, you can disable extra keys in the settings"?

Let me know what you think.

**Comment by Validbit** [comment]
I noticed 'č' is missing in AOSP layout, will fix it immediately once i have your feedback (to implement remaining changes in one commit)

**Comment by Validbit** [comment]
Thank you for your quick and detailed responses :)

I want to apologize if the layouts are too far from physical czech layout.
For the symbols placement I took inspiration from QWERTZ & QWERTY default layouts (Czech, German, English). 

They haven't been tested with time, but were made so they're more comfortable to use often. If wanted I could write details on what decisions were made on the design and why, but it might take some time I don't have atm.

The only thing I'd point out about the original proposal is the placement/spacing of brackets. Other than that I like it but (simply) likely woudn't use it, and believe there are better options to lay down for other languages, even if not the one mentioned in #383)

----------------
> Is this in competition with the layout developped in https://github.com/Julow/Unexpected-Keyboard/pull/380 ?
> If you do not plan to use this layout, I'd prefer to take @frimdo's version, as it would have at least one user.
> Being compliant with AOSP is not a goal as this app is very different.

Because this keyboards is different than usual soft keyboards (and soft keyboards handle much differently than traditional physical (layouts)) I made the "AOSP layout" (named for people who used AOSP-like keyb. commonly) close to requests and usual layouts of keyboards that many used before. It's not meant to compete but make compromise (for more potential users) not creating more layouts. If one/few people will use it, the option to import custom layout is well there, and can be made more user-friendly (#383)

> Is this a standard layout ? If yes, I'm in favor of adding it as it is.

No, Currently "Czech" is symbol-reduced (to proposed/common) "Czech Multifunctional" which is a QWERTZ layout that is custom to have all Czech symbols, while being easier to control (daily) by gestures (but compared to traditional layouts has steeper learning curve). If we were to just delete layouts (for now, _Multifunctional_ and proposed #380 (default) would be now prefferable on my side.

> Otherwise, the "extra keys" feature should be improved instead.
> It is not complete and covers only the languages defined in method.xml.

In #383 I made the "Czech" layout default and set the extra keys to only relevant (if it answers the question :))

> × should be preserved
> It's currently in the numeric board. Do you need it often enough to place it on the main board ?
Because I'm not that good at mathematics to pursue it in any way (and thus use it minimally outside of programming) no, but as said before...
```
Basic symbols:
+ = plus/add
- = minus
/ = division/comparison 
***** = markdown-like italics, quick-reference, any symbol, ...
(thus it'd be beneficial to have it there, since it's a basic operation)
```

> It's possible to place some keys on a layout without them being added if they are not needed. This is done with the loc  prefix, see for example in the QWERTY layout [...]
From what I understand, this is limited to being togglable with accents within existing system (dead-keys extra-keys), 

Loosly said : One of the things I love about this keybord is the speed (and not having to delay) when typing characters that are not special (but accented). That renders accent keys (not accented-character keys) not needed in my case (and people I saw using it)

> "Accents" option.
Do you mean the in-app settings option?  If yes i was not reffering to this but to option in xml that hides the accents by default (meaningful in multifunctional layout or not to "clutter" keyboard for first-time users)

**Comment by Julow** [comment]
@Validbit The layouts you propose are well made but don't solve this issue. Let's talk about them later.

People have placed brackets in weird places in the past. Perhaps to stay close to standards ?
I am in favor of modifying the layouts that you use or have eard complaints about.

It's nice that your layout have a purpose and a name. I would not remove a layouts.

> In #383 I made the "Czech" layout default and set the extra keys to only relevant (if it answers the question :))

How does they compare to https://github.com/Julow/Unexpected-Keyboard/pull/380 ? It was proposed before and already have a user.

The basic math symbols are part of the symbols I suggest to add in every layouts. @fridmo's layout have them. They are also part of the number board.

> [..] This is done with the loc prefix [..]
> From what I understand, this is limited to being togglable with accents within existing system (dead-keys extra-keys),

This is working in combination with extra-keys, which are not limited to accents or dead keys. In the QWERTY layout, it's used for many accents as well as `€₪ß£`.

> Loosly said : One of the things I love about this keybord is the speed (and not having to delay) when typing characters that are not special (but accented).

This is exactly why a less cluttered layout is important.

> If yes i was not reffering to this but to option in xml that hides the accents by default (meaningful in multifunctional layout or not to "clutter" keyboard for first-time users)

Do you mean keys that are present on the layout but invisible ?
Or do you mean keys that are placed on the layout (with the `loc ` prefix) but enabled only by users that want them (with the "add keys to the keyboard" option) ?

**Comment by Julow** [comment]
@Validbit Here's the last decision: https://github.com/Julow/Unexpected-Keyboard/pull/380 will be merged. The new layout will become the new default for Czech, the previous layout that you use will stay in the app under the name "multifunctional".

The only purpose of this discussion is now to understand how the many symbols you need could be fitted into the keyboard.

Some of my thoughts:

- Other layouts are not full of symbols. Free space increases readability and is needed for the "extra keys".

- We cannot have a "multifunctional" variant of every layouts but other people might want all these symbols.

- The app currently have places for unusual symbols: the number board, the greek&math board, the Fn layer.

- The compose key is a promising idea to have thousands more symbols: https://github.com/Julow/Unexpected-Keyboard/issues/204

By moving you out of a special case (a complex layout), if you agree, we can make the app better for everyone. Thanks for your help.

**Comment by Validbit** [comment]
> By moving you out of a special case (a complex layout), if you agree, we can make the app better for everyone

Of course. That was the goal of this discussion for me - developed from decluttering layouts (permitting official (ported) only) against custom practical layouts that were meant to be utilized, until a better solution is discussed, decided and implemented.

The compose key sounds promising, but I'd say the existing system of inputting these characters is better*, since (compared to solving the want for fast special keys input (like with AltGr)) it could be slower to remember (for those that aren't already well accessible now and for those who aren't used to them of course)

_*Implementing it would still benefit most so I'm up for it too_

What I'd suggest considering is
- **text expansion** triggered by special character 
     - (with togglable sets - importable/exportable with few **default ones** (e.g. defaults made (named) by use case: music (b), math (x) + latex, html, unicode, emoji-text, ctrlcode)
-  variant toggle (#383 comment)
     - since it's faster over time (what I'm trying to say repeatedly - no keyboard shortcut needed)
     - `off-topic: if swipe typing was as reliable (at least in non-english) many people would prefer it most of the time`


> People have placed brackets in weird places in the past. Perhaps to stay close to standards ?

After testing proposed and created AOSP layout I'd say yes but (universally) only if the direction is arbitrary - see images

_In the original and updated layout the bracket were put so that (gesture) typing is more intuitive (brackets are opposing and not too close, to begin and end by swiping like if you wanted to enclose the content (in brackets). The standart placement is also great but, imo, `only if the direction is arbitrary`_

_In AOSP(-like) they're close to one-handed (right-handed, as typical))_

![IMG_20230627_134929](https://github.com/Julow/Unexpected-Keyboard/assets/27623030/2dc1423e-9d62-4e53-921d-751360c706ab)

![IMG_20230627_135555](https://github.com/Julow/Unexpected-Keyboard/assets/27623030/974e47ac-3416-4942-8abf-e6b2ebe2371c)

![Screenshot_2023-06-27-13-45-45-82_3822ac0e29cfec8970059bb21d01f05e](https://github.com/Julow/Unexpected-Keyboard/assets/27623030/7be3c30e-2492-4aae-8edd-1d2de1c06096)

> Do you mean keys that are present on the layout but invisible ?
> Or do you mean keys that are placed on the layout (with the loc  prefix)
(placed on layout, loc)

> Free space increases readability and is needed for the "extra keys"

I agree but the thing is, for those who don't use dead keys for accents and are faster with it, not many extra keys are needed/possible. And (depending on quality of the layout, I'm not saying this layout is HQ) despite the higher learning curve (doesn't take much time spent) one can type things he couldn't (because it took longer or due to lack of knowledge/info about it) instead of wording it. 

As said before there are better options than fixed (baked, full) layouts (with symbols) but until the're developed and implemented that's why I'm using and developing the layout (I can't say how many symbols I need since I use these (all) - it depends on the case

> Here (without multiple layouts) it's still good, but... as some of us use our phones more and more for organisation, writing (email/text/document/blog/KB), school, programming (learning, queries) and mass communication on-the-go, there are simply more use-cases that "ask for" effortless typing.
> 
> Example uses of quick-access special symbols:
> 
> * communication/document shortcuts (Δ = change, ø = diameter of pizza, ~ = estimation, § = reference/law)
> * text-expansion (custom)
> * hobby/edu communication ( ♭, ÷)
> * passwords
> * systematic writing (blog, KB, wiki, ..) (×. `, •)
> * school (rest of symbols in current Multifunctional (- arrows))
> * communication itself (multiple languages, understanding without resorting to english)
> * language learning (multiple languages)
 [↩](#user-content-fnref-2-d0e36d18ad4e4e7fa4f1c273cbb4ce52)

**Comment by Julow** [comment]
- Text expansion is something that I plan to add on top of word suggestion/correction, that doesn't exist in the app currently. That would be called "custom dictionnary".
  Contribution in this direction is wholeheartedly welcome.

- Variant toggle is something to try. Would you have time to design such a feature ?

- I do not want to impose too much rules on layouts so I won't make general changes for brackets. Feel free to make changes to the layout you use.
  What do you mean by this ?:
  > only if the direction is arbitrary

  Also feel free to send PR to add more symbols to the various board/modifiers. (like you did in https://github.com/Julow/Unexpected-Keyboard/pull/383)

- There will surely be keyboard features requiring space for new keys (on the top of my head: voice typing shortcut, the compose key, editing keys, multilingual users).
  "Extra keys" is the main mechanism to allow typing in several languages.

- I do not believe that all the symbols you mention should be on the main board. First, they seem rarely used, second, some are already on the keyboard currently.

- I closed https://github.com/Julow/Unexpected-Keyboard/pull/383 for the layouts but the translations and the Fn combinations would be nice addition. Could you open a new PR ?

**Comment by Validbit** [comment]
> Variant toggle is something to try. Would you have time to design such a feature

I have little experience with this (I usually translate or edit xml-like structures, if) but I'll look into it next month. Still a proper text expansion (system - e.g. people could add/export own/others' sets) is more important/precursor (variants could e.g. use similar approach in choosing the sets, but that's just a rough idea)

> only if direction is arbitrary
so you can swipe both left & right to input it (as seen on the images, single assigned symbol is enough but there has to be space (unassigned symbols)). That way (i think) it's much easier to not make mistake (no matter user's typing preferences)

> I do not believe that all the symbols you mention should > be on the main board. First, they seem rarely used, > second, some are already on the keyboard currently.

As I later previously accepted the proposed layout (without these symbols) the ability to type them fast with minimal/no switching when needed (i.e. no combination/compose; but direct, or text expansion, etc.) is desired

I'll look into it when I have more time :) Should you release a new version soon, should I create a pull req. for translation, or is it already implemented? (it should be just copying the new language file)

**Comment by Validbit** [comment]
PR created.

I'd like to ask (curiosity)... 
1. How would word suggestion/correction impact performace/battery-usage? (I've seen many keybords gradually got worse/lagging after gradually developing prediction/tool-bar)

I'm not against implementing it (and understand is easier to make expansion through it (checking every character, instead of e.g. adding function call to special key(s)) but this keybord is one the few that allows people to type confidently since word-correction won't make mistakes (accents, different words, shortcuts, project/made-up words, etc.) but they still are able to type without delay/checking (+ better muscle memory once they're used to it, if)

2. Would it be togglable between character-text-expansion and word-correction/hinting?

Thank you for your great work.

**Comment by Julow** [comment]
Sorry for the slow reply.
Word suggestion/correction would have a very little impact. Keeping the app lightweight and fast is a goal. Word correction/hinting is what I have in mind.
I'm not currently working on it, though.

**Comment by Julow** [comment]
Is this issue still relevant ? The Czech layout is no longer cluttered and the keyboard can now type more symbols.

**Comment by frimdo** [comment]
I am using the uncluttered czech layout and have no issues with it. I am happy with the result. I am closing this issue as fixed :) .

**Comment by Julow** [comment]
Thanks :) Feel free to modify the layout if needed.
